### ISSUE TITLE
import NAIP data into a postgres database

### ISSUE THREAD

**Issue by andrewljohnson** [trailbehind/DeepOSM#23]
Putting the data in Postgres seems like a good mid-game/end-game move. Do this after we put up deeposm.org, want to scale, and/or want to provide a place for researchers to run arbitrary experiments.

Benefits include: 
- make rotating tiles easier (issue #24) - current pipeline could me modified too, but more hacky
- maybe easier to do bounding box queries, than if data is cached in a non-relational way from NAIPs to disk
- enable an API that would allow for more arbitrary training data, in less disk space

**Comment by silberman** [comment]
...we provide a function where you could say "give me 40,000 random tiles from within these long,lat bounding boxes, and label them using these labeller functions I want to try", and that would be fast.

(A "labeller" function being something that takes a long,lat bounding box and returns some numpy array, with simple ones that return 1:1 arrays of one of the RGBI bands, or more complex ones like the has_center_road and its various permutations, or ones that map 64x64 to 4x4 binary has-road, or has-tennis, etc)

We could still cache it if we want, (as is being discussed in https://github.com/trailbehind/DeepOSM/issues/30 ), though I think we can ditch all the NAIP-specific details, and just save to NetCDF the arrays that are going straight into tensorflow, plus some metadata about the experiment if we want.

**Comment by zain** [comment]
@silberman: what format would you use for the postgis database? Just whatever osm2pgsql or imposm spits out?

**Comment by silberman** [comment]
Not sure, and very open to suggestions.  This will be my first time using postgis.  Planning to spend some time on this over the weekend, so I'll know a lot more then.

**Comment by andrewljohnson** [comment]
How about:

1) skip postgis
2) use the Overpass API, hosted by some friendly Germans:
http://overpass-api.de/
3) write a method to query Overpass for roads for a bounding box (i.e. NAIP
bounding box)
4) throw away my PBF munging code
5) enjoy the ease of not having OSM data locally

**Comment by zain** [comment]
I don't know anything about Overpass but we just want an API for OSM data,
Mapzen also has a great vector tiles API (that's what I used for my initial
experiments). However, getting OSM data already takes a while and adding a
bunch more HTTP queries would make it longer. I think a
persistent postgis database would make things faster instead of slower.

It's also nice because that means we can spin up/down GPU instances more
quickly -- don't need to download a bunch of OSM data each time we start
one.

Andrew, out of curiosity, what do you see as the pains of local OSM data?

**Comment by andrewljohnson** [comment]
1) The only real pain I see is setting it up... and I fear there is some
sort of maintenance required but can't quantify that fear.

2) I think doing bounding box queries against Overpass will be really
fast... waaaaaay faster than the current local PBF parser. It will also
remove the time to download the PBF files and extract the ways. Instead,
you do one bounding box query per NAIP I think, which should take a second
or so.

Also, you can end up caching the overpass queries and/or way bitmaps that
get generated.

**Comment by zain** [comment]
How much parsing of the PBFs are we doing? Are they state-level PBFs and we do a full parse for each bounding box?

**Comment by andrewljohnson** [comment]
The current parse works like this:

* open N state-level PBFs
* make a giant list of all the ways involved in the analysis (i.e. highways, but later building too, or cycleways/footways)
* for each NAIP, loop over list and isolate ways that intersect the NAIP, and convert to bitmap

**Comment by zain** [comment]
Ok. opened up #39 to set up an Overpass server to try it out.

**Comment by andrewljohnson** [comment]
merging with other infrastructure issues
